### ISSUE TITLE
Vine: Manager Stuck if Workers with VINE_TEMP Time Out

### ISSUE THREAD

**Issue by btovar** [cooperative-computing-lab/cctools#3531]
One of my runs got into an endless loop because temporary results were lost as workers idled out. By the time those results were regenerated, other workers had idled out, etc.   I think it is expected for workers not to idle out when they hold VINE_TEMP files as they are not really idle (i.e., they are part of the distributed cache).

**Comment by dthain** [comment]
Hmm, this is a tricky problem with competing goals:
- Workers can't stick around forever: we need a garbage collection mechanism for when the manager just disappears.  It doesn't have to be 15 minutes, but it can't be infinite either.
- Temporary files permit a manager to take a risk by leaving a file in place, and then coming back to it later.  If the bet fails, then the manager has to pay the price of regenerating it.  If we have to regenerate too many files, then the manager doesn't make progress.
- By design, the worker doesn't really know the intentions of the manager.  And while it does (sort of) know which files are temporaries, it doesn't have a full understanding of how they will be used, how many times they will be used, etc...

**Comment by dthain** [comment]
@colinthomas-z80 your thoughts?

**Comment by btovar** [comment]
It may be helpful to divide the problem in two:

A. How to deal with temp files when the worker is still connected to the manager and idles out.
B. How to deal with the cache when the worker disconnects from the manager for any reason.

This issue is just A. B is more complicated but only matters for files cached forever. In any other case it wouldn't be surprising for the worker to clean non-forever files on disconnection.

Automatically converting to non-temp files may not work, as the manager may not have the local disk necessary to host the files, or the workers live in some cloud and the manager does not want to pay for the transfers.

My ideal solution would be for the manager to communicate to the worker not to idle out if it has temp files. When files are no longer needed, the manager application can call `vine_file_remove` and let the worker idle out when needed. I'm ok with the manager sending a temp flag when transmitting the file information.

Currently I have to set the timeout for 1h to be on the safe side, but this just like saying "don't garbage collect", but this wastes tons of resources.

**Comment by colinthomas-z80** [comment]
Could the worker ask the manager permission to time out? If the manager doesn’t exist it will proceed as normal, but if the manager is still around it may respond and tell the worker to stick around, or transfer the temporary data back before exiting

**Comment by colinthomas-z80** [comment]
Or the manager could just send keep-alives to workers with important data

**Comment by dthain** [comment]
I like the second idea better (manager sends keep-alives to workers with temp files) b/c it is in keeping with the general philosophy that workers have mechanisms, but the manager is in charge of general policy. 

Now, to clarify terminology a little bit:

The worker has two different timeouts:

1 - The `idle_timeout` counts the time since a task or other major action has been received, while connected to a manager.  The worker will disconnect if the manager has not given it any worker, even while connected.  I note that `handle_manager` calls `reset_idle_timer` on receipt of some but not all messages.  We should probably look at this carefully to make sure that it is still up to date:
https://github.com/cooperative-computing-lab/cctools/blob/master/taskvine/src/worker/vine_worker.c#L1054

2 - The `connect_timeout` counts the time since the last successful connection, and then bails out if no connection can be made.


So, 1 handles the case of a manager with more workers than it actually needs (which also allows for workers to load-balance across multiple managers).  While 2 handles the case of a totally absent manager.

**Comment by dthain** [comment]
Also note that the manager already sends periodic keepalives just to make sure that the connection is active:
https://github.com/cooperative-computing-lab/cctools/blob/master/taskvine/src/manager/vine_manager.c#L3221C11-L3221C11

**Comment by btovar** [comment]
I don't like the idea of the timeouts, as setting timeouts is hard! And more when we can determine logically what should happen. The manager is still in charge of the policy, that is, don't disconnect while you have this file. It is the same as don't disconnect while you have a task.

Another thing I don't like is that the timeout of the worker has to be somewhat coordinated with the periods of the keep-alives, which means a user has to be aware of this internal knob of taskvine.  If the worker sets a timeout too low, or we modify the keepalive interval inside taskvine, it will miss the keep alive.

The current manager keepalive has a different function. It is asking whether the worker is still there, rather than telling the worker not to idle out. (It is a "please send me a keep alive".)

**Comment by colinthomas-z80** [comment]
How would we handle workers with no timeouts when the manager exits?

**Comment by btovar** [comment]
The worker still has the timeout for giving up when no manager is found. This timeout is independent of any timeout in the manager.

Also, that's why I'd like to divide this problem into what to do with workers currently connected that have files we are interested in, and what to do whit the files on connection/disconnection.

I'd like to avoid the timeout solution because timeout interactions are much harder to debug than logical conditions, and they may be a little brittle against changes in the manager operation.

**Comment by dthain** [comment]
Timeouts are ugly, true.

Let me also point out that the problem is broader than just temporary files.

In Work Queue, the data storage was incidental to running a task.  If no task was sent to a worker, it wasn't being used.

But in TaskVine, the worker still provides value, even if it isn't running a task.  It may be storing temporary files, yes.  But it may also be serving as a source of third-party transfers, keeping serverless libraries warm, etc.

What would be the consequences of removing the idle timeout entirely?  (We would still have the connect timeout, of course.)

**Comment by btovar** [comment]
I think we want the workers to idle-out at the ramp-down of workflows with long tails.  Also without the idle timeout, the factory becomes less effective at managing resources. That reminds me, the factory pruning mechanism likely is able to remove workers as soon as they complete a task that produced a VINE_TEMP.

What about this: a worker can idle-out if:
- It is not running a task.
- It is not have pending results to send back.
- It does not have temporary files in its cache, (or does not have a temp file with VINE_DONT_IDLE_OUT).
- It does not have a library marked as VINE_DONT_IDLE_OUT.

In other words, idles-out if it currently does not provide any value.

My current solution of setting the timeout to a very large value is equivalent to removing the idle-timeout, and my condor priority is taking a hit just waiting for those last accumulation tasks, ha. The manager has been removing temp files as soon as they are no longer needed, so the worker could know if it is providing value to the manager.

**Comment by dthain** [comment]
The more complicated these conditions get, the more I am convinced that the manager should make the decisions.  :)

So, if ramp-down is the only scenario in which we want workers to disconnect from an otherwise healthy manager, then it seems to me that the manager knows when ramp down is in effect, and can start to remove workers that meet its criteria.  Much like we do `shutdown_slow_workers`.

This is also gives the manager an opportunity to reason about the total resources needed.  For example, during ramp-down, you might want to maintain N idle workers, just in case some of the existing ones fail.

**Comment by btovar** [comment]
I think that if we were talking about Work Queue, I'd agree with you. In Work Queue we did not disconnect workers with tasks or with results, and the main abstraction was the task. In TaskVine the files are an explicit part of the abstraction, and the power comes from the distributed cache. In some sense, tasks are just a medium to populate that cache (otherwise, we would just kept using Work Queue). Thus, if the distributed cache is important, then the mechanisms of the worker should reflect that.

About detecting ramp-down, one problem is that the application logically may think that it is in ramp-down, but the actual execution may not be. For example, the manager may think that all tasks have completed because they returned via wait, but in reality taskvine is re-executing tons of tasks because some mass eviction. Thus, the manager may think it can allow workers to idle-out when in reality it needs as many workers as before.

I don't like all these hidden interactions. It is much easier to say to a prospective user: "if a worker has a VINE_TEMP file, it won't idle-out", rather than, "if a worker is in ramp-down mode, which depending on your application, and the number of tasks, and the number of workers, and the number of waiting tasks, etc.". Ramp-down optimizations are sometimes needed, but I feel they should be a power-user tool, that is, the distributed cache should be easily accessible to new taskvine users.  Temp results disappearing from the cache because of timeouts is unexpected.

**Comment by dthain** [comment]
Actually, I think we are not that far apart.

I see your point about ramp down, it's not as simple as I thought.

I think we agree on the conditions for idle timeout -- no tasks running/pending, no temporary files -- then the only remaining question is, what party is responsible for making that decision?

Is it better for the manager to periodically evaluate the known workers and kick them off, or for workers to make that judgement independently and disconnect when they decide to?

**Comment by btovar** [comment]
It seems more natural for the worker to make that decision. The manager has control of the worker resources when the worker connect to it, but the manager is not in total control of the worker's lifetime (the factory, or some other process is). For long lived workers, if they are released by the manager we do not want them to try to connect again and again. (And we can't tell the worker not to connect to some port:host, as it could be that another manager has appeared on that same address.)

If the manager is the one giving permission to idle-out, we need to book-keep permissions and revocations every time the state of the assets at the worker change. The worker already has all the information to make this determination, and we wouldn't need the extra bookkeeping at the manager.

**Comment by dthain** [comment]
While looking at Jin's problem, I realized that you are right, it has to be the worker that makes the decision b/c when the manager is overloaded, it doesn't have the *time* to disconnect unused workers.

 So now the problem is to figure out how to get the relevant info.  The worker has the *name* of each file, which is derived from the type and the metadata, but not the metadata itself...

**Comment by dthain** [comment]
So clearly this has opened up a large discussion of how to perform data management, which is great.  But, we will need a *simple* solution worked up in short order so that users of TopEFT are not wasting large numbers of workers as their computation ramps down.  I want to come up with a meaningful solution that can be implemented in days, not weeks, and leaves open the ability to refine the solution later.

So I propose the following:

The basic architectural concept is that the manager should periodically iterate over all workers and direct them to shutdown if they are not contributing to the workflow, either by running jobs or storing data.  (And this is the manager's job b/c it knows best the overall state of the computation.)

Specifically, the manager should remove a worker if:
- The worker has been idle for 15 minutes (configurable)
- And it is not storing any TEMP files with a replica count of one.

At the same time, we should remove the worker's independent decision to disconnect from the manager.  It does not have the scope of information needed to make that decision.  (e.g. even if it knows that it has a TEMP file, it has no idea how many copies there are.)

@btovar @colinthomas-z80 what do you think?

**Comment by btovar** [comment]
Sounds good to me!

**Comment by colinthomas-z80** [comment]
Sounds good.

**Comment by dthain** [comment]
Ok @colinthomas-z80 this one is yours.  :)

**Comment by colinthomas-z80** [comment]
Would it be too much of a deviation from the proposed implementation to just change the original worker timeout to a timeout request, and let the manager decide to allow it or not? The interaction can be asynchronous. 

This way the manager does not need to spend time iterating through workers and maintaining timestamps, and the timeout configuration will continue to work the same way. 

If the manager becomes responsible for keeping timeout info, there could be some inefficiency with tasks that run longer than the timeout period. The manager would enter a state thinking the worker might be idle until it realizes it is running a task. I think the worker has the best understanding of its real idle time.

**Comment by colinthomas-z80** [comment]
Well after reading ben's comments in #3538 again I can see the argument for the manager side. Might be better off for future items that way.

**Comment by btovar** [comment]
If the manager rejects the request, does the worker resets its timer? Does the worker gets flagged somehow so that the manager can later drain the worker and release the worker? Does the worker assume that no response means no?

**Comment by dthain** [comment]
These are all good Qs.  @colinthomas-z80 please sketch out how you think this should work before coding it up...

**Comment by colinthomas-z80** [comment]
The timer can get reset each time it sends out a request before any response is received. There may not be a need for the manager to even respond "no". 

If the manager decides the worker should be removed it would be flagged for drain so no more tasks get sent out. Then it can be removed through the drain routine on the manager's time.

**Comment by colinthomas-z80** [comment]
I have gone through a number of workflows testing different timeout configurations. I am pretty confident that #3555 resolves this issue for now.

**Comment by dthain** [comment]
Sounds good.
